lightning/src/ln/functional_test_utils.rs

```diff
@@ -1191,24 +1191,85 @@ macro_rules! get_route_and_payment_hash {
 	}}
 }
 
+pub struct PaymentForwardedFailedConditions {
+	pub expected_pending_htlcs_forwardable: bool,
+	pub expected_payment_forwarding_failed: Option<u32>,
+}
+
+impl PaymentForwardedFailedConditions {
+	pub fn new() -> Self {
+		Self {
+			expected_pending_htlcs_forwardable: false,
+			expected_payment_forwarding_failed: None,
+		}
+	}
+
+	pub fn pending_htlcs_forwardable(mut self) -> Self {
+		self.expected_pending_htlcs_forwardable = true;
+		self
+	}
+
+	pub fn payment_forwarding_failed(mut self) -> Self {
+		self.expected_payment_forwarding_failed = Some(1);
+		self
+	}
+
+	pub fn payment_forwarding_failed_with_count(mut self, count: u32) -> Self {
+		self.expected_payment_forwarding_failed = Some(count);
+		self
+	}
+}
+
 #[macro_export]
-/// Clears (and ignores) a PendingHTLCsForwardable event
-macro_rules! expect_pending_htlcs_forwardable_ignore {
-	($node: expr) => {{
+macro_rules! expect_pending_htlcs_forwardable_conditions {
```

Comment:
From this line to 1284, I'm not entirely sure if this is the best way to name and compose macros. definitely open to suggestions for improvement here.